repose-aggregator/src/docs/asciidoc/filters/cors.adoc

```diff
@@ -290,64 +337,57 @@ the specified URI will be treated as a Java regular expression.
+A web browser may choose to skip sending a CORS Preflight Request if the HTTP method is GET, HEAD, or POST, the request headers do not include anything other than `Accept`, `Accept-Language`, and `Content-Language`, and the request does not require any cookies, HTTP authentication, nor use of any client-side SSL certificates.
+Otherwise, the web browser must make a CORS Preflight Request.
+For example, if your origin service requires an `X-Auth-Token` header, the web browser will always send a CORS Preflight Request before sending the CORS Actual Request.
 
-* HTTP method is GET, HEAD, or POST
-* Headers do not include anything other than `Accept`, `Accept-Language`, `Content-Language`, and `Content-Type`
-(assuming the values are `application/x-www-form-urlencoded`, `multipart/form-data`, or `text/plain`).
+==== Origin Header
+Some web browsers (e.g., Chrome and Safari) will send the `Origin` header for same-origin (i.e., non-CORS) requests in addition to CORS requests which is technically allowed under https://tools.ietf.org/html/rfc6454#section-7.3[RFC 6454].
+This is typically not a concern for servers handling CORS because the unnecessary inclusion of CORS headers will be ignored by the client if they are not needed to process the response.
+Because Repose has the added ability to reject requests with unapproved origins, additional logic is required to differentiate between CORS requests and same-origin requests when the `Origin` header is present.
 
-NOTE: If your origin service requires a special header (e.g. `X-Auth-Token`),
-a Preflight Request will always be sent for CORS requests.
+Requests are considered same-origin requests when the `Origin` header matches the original `Host` header set by the client according to the comparison rules in https://tools.ietf.org/html/rfc6454#section-5[RFC 6454 Section 5].
+If the `X-Forwarded-Host` header is present in the request, the first value will be used as the host.
+If the header is not present or it cannot be parsed as a URI (when the URI scheme is prepended to it), then the `Host` header will be used instead.
+This check is not performed for CORS Preflight Requests since web browsers should not be sending a CORS preflight header for a same-origin request.
 
-WARNING: CORS is not a security feature.
-It is a mechanism for informing clients (e.g. web browsers) of conditions when client-side security can be slightly relaxed in certain circumstances.
-That is, the security lies completely within the client.
-Simply leaving out the `Origin` header in the request completely bypasses the CORS spec (and thus this filter).
-You should continue securing your API in other ways using proper authentication and authorization.
+[CAUTION]
+====
+If a proxy server sitting between the client and Repose rewrites the `Host` without updating the `X-Forwarded-Host` header with the original value,
+Repose will not be able to correctly identify same-origin requests coming from that client.
```

Comment:
Bah!  I had been putting a new line after the comma before, but in this changeset, I started undoing that since I think we decided one sentence per line, no more, no less.  To make it easy, I went line by line looking for lines starting with lowercase letters.  I missed this line because it started with "Repose".  I will put it on the previous line.